### PR TITLE
fix chunk optimization

### PR DESCRIPTION
**Pull request by sokra**
before it just merged all stuff into one chunk when over the chunk limit

### CODE REVIEW

File: crates/turbopack-core/src/chunk/optimize.rs [hunk 1]
```diff
@@ -2,18 +2,22 @@ use std::{cell::RefCell, mem::take, rc::Rc};
```
Comment (ForsakenHarmony) [inline crates/turbopack-core/src/chunk/optimize.rs:2]: would be nice to add some module level docs that give a high level overview of what's happening here

File: crates/turbopack-ecmascript/src/chunk/optimize.rs [hunk 2]
```diff
@@ -1,4 +1,7 @@
-use std::{cmp::Ordering, collections::HashSet};
+use std::{
```
Comment (ForsakenHarmony) [inline crates/turbopack-ecmascript/src/chunk/optimize.rs:1]: would be nice to add some module level docs that give a high level overview of what's happening here (same as above)

**Review by ForsakenHarmony — APPROVED** [review]

File: crates/turbopack-ecmascript/src/chunk/optimize.rs [hunk 3]
```diff
@@ -269,11 +272,69 @@ async fn merge_duplicated_and_contained(
...
+    target_count: usize,
+) -> Result<Vec<EcmascriptChunkVc>> {
+    let mut remaining = chunks.len();
+    let mut chunks = chunks.into_iter().enumerate().collect::<Vec<_>>();
+    chunks.sort_by_key(|&(i, (_, k))| (k, i));
+    let mut chunks_by_source = chunks
+        .group_by(|(_, (_, a)), (_, (_, b))| a == b)
+        .map(|slice| {
+            (
+                slice[0].1 .1,
```
Comment (jridgewell) [inline crates/turbopack-ecmascript/src/chunk/optimize.rs:286]: Weird space
```suggestion
                slice[0].1.1,
```

File: crates/turbopack-ecmascript/src/chunk/optimize.rs [hunk 4]
```diff
@@ -269,11 +272,69 @@ async fn merge_duplicated_and_contained(
     Ok(())
 }
 
-async fn merge_by_size(chunks: &[EcmascriptChunkVc]) -> Result<Vec<EcmascriptChunkVc>> {
+async fn merge_to_limit(
+    chunks: Vec<(EcmascriptChunkVc, Option<ChunksVc>)>,
+    target_count: usize,
+) -> Result<Vec<EcmascriptChunkVc>> {
+    let mut remaining = chunks.len();
+    let mut chunks = chunks.into_iter().enumerate().collect::<Vec<_>>();
```
Comment (jridgewell) [inline crates/turbopack-ecmascript/src/chunk/optimize.rs:280]: I think the enumerate (and the nested tuple complications below) is unnecessary? I assume you're doing it to ensure a stable sorting of the chunks, but Rust's sort it stable already.

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: crates/turbopack-core/src/chunk/optimize.rs [hunk 5]
```diff
@@ -169,6 +197,19 @@ impl ContainmentTree {
...
 
+        fn skip_unnessary_nodes(trees: &mut IndexMap<FileSystemPathVc, Rc<RefCell<Node>>>) {
+            for tree in trees.values_mut() {
+                let mut tree = tree.borrow_mut();
+                if tree.chunks.len() == 0 && tree.children.len() == 1 {
+                    let child = tree.children.pop().unwrap();
+                    let mut child = child.borrow_mut();
+                    tree.path = child.path;
+                    tree.chunks.append(&mut child.chunks);
+                    tree.children.append(&mut child.children);
```
[the remark this answers is not in the document]
Comment (sokra) [reply crates/turbopack-core/src/chunk/optimize.rs:208]: Good point

File: crates/turbopack-ecmascript/src/chunk/optimize.rs (hunk 3 above)
Comment (sokra) [reply crates/turbopack-ecmascript/src/chunk/optimize.rs:286]: cargo fmt adds this. I guess it should not be a number literal

File: crates/turbopack-core/src/chunk/optimize.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (sokra) [reply crates/turbopack-core/src/chunk/optimize.rs:208]: Seems like the AI wrote that correctly^^

File: crates/turbopack-ecmascript/src/chunk/mod.rs [hunk 6]
```diff
@@ -126,6 +126,7 @@ impl EcmascriptChunkVc {
             .resolve()
             .await?;
         for path in paths {
+            let path = path.resolve().await?;
```
[the remark this answers is not in the document]
Comment (sokra) [reply crates/turbopack-ecmascript/src/chunk/mod.rs:129]: I don't see how this is possible...

File: crates/turbopack-core/src/chunk/optimize.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jridgewell) [reply crates/turbopack-core/src/chunk/optimize.rs:208]: Ohh, I missed the `pop()` call.

**Review by jridgewell — APPROVED** [review]

**Comment by ForsakenHarmony** [comment]
@sokra why did you ignore the clippy warnings?

**Pull request merged**